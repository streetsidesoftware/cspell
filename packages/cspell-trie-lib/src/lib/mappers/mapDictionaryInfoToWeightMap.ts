import { createWeightMap, WeightMap } from '../distance/weightedMaps';
import type { DictionaryInformation, HunspellCosts, HunspellInformation } from '../models/DictionaryInformation';
import type { SuggestionCostMapDef } from '../models/suggestionCostsDef';
import { isDefined } from '../utils/util';

type Costs = Required<HunspellCosts>;

export function dictionaryInformationToWeightMap(dictInfo: DictionaryInformation): WeightMap {
    const defsEC = dictInfo.suggestionEditCosts || [];
    const defsHI = dictInfo.hunspellInformation
        ? hunspellInformationToSuggestionCostDef(dictInfo.hunspellInformation)
        : [];
    return createWeightMap(...defsEC, ...defsHI);
}

export function hunspellInformationToSuggestionCostDef(hunInfo: HunspellInformation): SuggestionCostMapDef[] {
    const defs: SuggestionCostMapDef[] = [];
    const costs = calcCosts(hunInfo.costs);

    function parseAff(aff: string, costs: Costs) {
        // cspell:ignore OCONV
        const regSupportedAff = /^(?:MAP|KEY|TRY|NO-TRY|ICONV|OCONV|REP)\s/;
        const rejectAff = /^(?:MAP|KEY|TRY|ICONV|OCONV|REP)\s+\d+$/;

        const lines = aff
            .split('\n')
            .map((a) => a.replace(/#.*/, ''))
            .map((a) => a.trim())
            .filter((a) => regSupportedAff.test(a))
            .filter((a) => !rejectAff.test(a));
        lines.forEach((line) => {
            [affMap, affNoTry, affRepConv, affTry, affTryFirstCharacterReplace, affKey]
                .map((fn) => fn(line, costs))
                .filter(isDefined)
                .forEach((def) => defs.push(def));
        });
    }

    parseAff(hunInfo.aff, costs);

    return defs;
}

function calcCosts(costs: HunspellCosts = {}): Costs {
    const {
        firstLetterPenalty = 4,
        ioConvertCost = 30,
        keyboardCost = 94,
        mapCost = 25,
        replaceCosts = 75,
        tryCharCost = 95,
    } = costs;
    const c: Costs = { tryCharCost, keyboardCost, mapCost, ioConvertCost, replaceCosts, firstLetterPenalty };
    return c;
}

const regExpMap = /^(?:MAP)\s+(\S+)$/;

function affMap(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpMap);
    if (!m) return undefined;

    const map = m[1];
    const cost = costs.mapCost;

    return {
        map,
        replace: cost,
        swap: cost,
    };
}

const regExpTry = /^(?:TRY)\s+(\S+)$/;

function affTry(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpTry);
    if (!m) return undefined;

    const map = m[1];
    const cost = costs.tryCharCost;

    return {
        map,
        insDel: cost,
        replace: cost,
        swap: cost,
    };
}

function affTryFirstCharacterReplace(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpTry);
    if (!m) return undefined;

    const map = [...split(m[1])].map((char) => `(^${char})`).join('');
    const cost = costs.tryCharCost;
    const penalty = costs.firstLetterPenalty;

    return {
        map,
        replace: cost - penalty,
        penalty,
    };
}

const regExpNoTry = /^NO-TRY\s+(\S+)$/;

function affNoTry(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpNoTry);
    if (!m) return undefined;

    const map = m[1];

    return {
        map,
        insDel: Math.max(100 - costs.tryCharCost, 0),
        penalty: 100 + costs.tryCharCost,
    };
}

// cspell:ignore conv
const regExpRepConv = /^(?:REP|(?:I|O)CONV)\s+(\S+)\s+(\S+)$/;

function affRepConv(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpRepConv);
    if (!m) return undefined;

    const cost = line.startsWith('REP') ? costs.replaceCosts : costs.ioConvertCost;
    const from = m[1];
    let into = m[2];
    into = into.replace(/^0$/, '');

    if (from.startsWith('^') && !into.startsWith('^')) {
        into = '^' + into;
    }

    if (from.endsWith('$') && !into.endsWith('$')) {
        into = into + '$';
    }

    return {
        map: `(${from})(${into})`,
        replace: cost,
    };
}

const regExpKey = /^(?:KEY)\s+(\S+)$/;

function affKey(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpKey);
    if (!m) return undefined;

    const kbd = m[1];

    const pairs = [...split(kbd)]
        .map(reducer((p, v) => ({ a: p.b, b: v }), { a: '|', b: '|' }))
        .filter((ab) => ab.a !== '|' && ab.b !== '|')
        .map((ab) => ab.a + ab.b);

    const map = pairs.join('|');
    const cost = costs.keyboardCost;

    return {
        map,
        replace: cost,
        swap: cost,
    };
}

function* split(map: string): Iterable<string> {
    let seq = '';
    let mode = 0;
    for (const char of map) {
        if (mode && char === ')') {
            yield seq;
            mode = 0;
            continue;
        }
        if (mode) {
            seq += char;
            continue;
        }
        if (char === '(') {
            mode = 1;
            seq = '';
            continue;
        }
        yield char;
    }
}

function reducer<T, U = T>(fn: (acc: U, val: T, i: number) => U, initialVal: U) {
    let acc = initialVal;
    return (val: T, i: number) => (acc = fn(acc, val, i));
}

// function pipe<T>(v: T, ...fns: ((v: T) => T)[]): T {
//     for (const fn of fns) {
//         v = fn(v);
//     }
//     return v;
// }

export const __testing__ = {
    affKey,
    affMap,
    affNoTry,
    affRepConv,
    affTry,
    affTryFirstCharacterReplace,
    split,
    calcCosts,
};
