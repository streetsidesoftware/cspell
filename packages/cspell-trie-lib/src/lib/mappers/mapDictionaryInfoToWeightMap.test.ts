import type { HunspellCosts, HunspellInformation } from '../models/DictionaryInformation';
import { hunspellInformationToSuggestionCostDef, __testing__ } from './mapDictionaryInfoToWeightMap';

// cspell:ignore conv OCONV
const { affMap, affRepConv, affNoTry, affTry, affKey, affTryFirstCharacterReplace, calcCosts } = __testing__;

const sampleAff = `
# comment.
# cspell:ignore eéèëê iíìïî
TRY abc

MAP 3
MAP eéèëê
MAP aáà
MAP iíìïî

OCONV 2
OCONV ĳ ij
OCONV Ĳ IJ

ICONV 9
ICONV áá aa
ICONV éé ee
ICONV íé ie
ICONV óó oo
ICONV úú uu
ICONV óé oe
ICONV ’ '
ICONV ij ĳ
ICONV IJ Ĳ

REP ^ß ss

NO-TRY -1234567890

# cspell:ignore asdfghjkl qwertyuiop zxcvbnm
KEY qwertyuiop|asdfghjkl|zxcvbnm
`;

describe('mapAffToWeightMap', () => {
    // cspell:ignore aàâä
    test.each`
        line                  | costs             | expected
        ${''}                 | ${{}}             | ${undefined}
        ${'MAP aàâäAÀÂÄ'}     | ${{}}             | ${{ map: 'aàâäAÀÂÄ', replace: 25, swap: 25 }}
        ${'MAP 😁😀😊😂🤣😬'} | ${{}}             | ${{ map: '😁😀😊😂🤣😬', replace: 25, swap: 25 }}
        ${'MAP aàâäAÀÂÄ'}     | ${{ mapCost: 1 }} | ${{ map: 'aàâäAÀÂÄ', replace: 1, swap: 1 }}
    `('affMap "$line" $costs', ({ line, costs, expected }) => {
        expect(affMap(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line                 | costs                            | expected
        ${''}                | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'}    | ${{}}                            | ${undefined}
        ${'NO-TRY aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY abc'}         | ${c({ mapCost: 1 })}             | ${{ map: 'abc', replace: 95, insDel: 95, swap: 95 }}
        ${'TRY abc'}         | ${c({ tryCharCost: 90 })}        | ${{ map: 'abc', replace: 90, insDel: 90, swap: 90 }}
        ${'TRY abc'}         | ${c({ firstLetterPenalty: 10 })} | ${{ map: 'abc', replace: 95, insDel: 95, swap: 95 }}
    `('affTry "$line" $costs', ({ line, costs, expected }) => {
        expect(affTry(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                            | expected
        ${''}             | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'NO-TRY abc'}   | ${c({ mapCost: 1 })}             | ${{ map: 'abc', insDel: 5, penalty: 195 }}
        ${'NO-TRY abc'}   | ${c({ tryCharCost: 90 })}        | ${{ map: 'abc', insDel: 10, penalty: 190 }}
        ${'NO-TRY abc'}   | ${c({ firstLetterPenalty: 10 })} | ${{ map: 'abc', insDel: 5, penalty: 195 }}
    `('affNoTry "$line" $costs', ({ line, costs, expected }) => {
        expect(affNoTry(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                            | expected
        ${''}             | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY abc'}      | ${c({ mapCost: 1 })}             | ${{ map: '(^a)(^b)(^c)', replace: 91, penalty: 4 }}
        ${'TRY abc'}      | ${c({ tryCharCost: 90 })}        | ${{ map: '(^a)(^b)(^c)', replace: 86, penalty: 4 }}
        ${'TRY abc'}      | ${c({ firstLetterPenalty: 10 })} | ${{ map: '(^a)(^b)(^c)', replace: 85, penalty: 10 }}
    `('affTryFirstCharacterReplace "$line" $costs', ({ line, costs, expected }) => {
        expect(affTryFirstCharacterReplace(line, calcCosts(costs))).toEqual(expected);
    });

    // cspell:ignore qwer zxcv
    test.each`
        line                          | costs                      | expected
        ${''}                         | ${{}}                      | ${undefined}
        ${'MAP aàâäAÀÂÄ'}             | ${{}}                      | ${undefined}
        ${'KEY qwer|asdf|zxcv'}       | ${c({ mapCost: 1 })}       | ${{ map: 'qw|we|er|as|sd|df|zx|xc|cv', replace: 94, swap: 94 }}
        ${'KEY qwer|a|asdf|zxcv'}     | ${c({ keyboardCost: 74 })} | ${{ map: 'qw|we|er|as|sd|df|zx|xc|cv', replace: 74, swap: 74 }}
        ${'KEY qwer春😁|a|asdf|zxcv'} | ${c()}                     | ${{ map: 'qw|we|er|r春|春😁|as|sd|df|zx|xc|cv', replace: 94, swap: 94 }}
        ${'KEY a😁b'}                 | ${c()}                     | ${{ map: 'a😁|😁b', replace: 94, swap: 94 }}
    `('affKey "$line" $costs', ({ line, costs, expected }) => {
        expect(affKey(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                    | expected
        ${''}             | ${{}}                    | ${undefined}
        ${'REP o oo'}     | ${{}}                    | ${{ map: '(o)(oo)', replace: 75 }}
        ${'REP ^a A'}     | ${{ mapCost: 1 }}        | ${{ map: '(^a)(^A)', replace: 75 }}
        ${'REP $ en$'}    | ${{ replaceCosts: 55 }}  | ${{ map: '($)(en$)', replace: 55 }}
        ${'REP ^af$ aff'} | ${{}}                    | ${{ map: '(^af$)(^aff$)', replace: 75 }}
        ${'REP ß ss'}     | ${{}}                    | ${{ map: '(ß)(ss)', replace: 75 }}
        ${'REP ß 0'}      | ${{}}                    | ${{ map: '(ß)()', replace: 75 }}
        ${'REP 25'}       | ${{}}                    | ${undefined}
        ${'ICONV 25'}     | ${{}}                    | ${undefined}
        ${'OCONV 25'}     | ${{}}                    | ${undefined}
        ${'ICONV áá aa'}  | ${{}}                    | ${{ map: '(áá)(aa)', replace: 30 }}
        ${'OCONV ss ß'}   | ${{ replaceCosts: 55 }}  | ${{ map: '(ss)(ß)', replace: 30 }}
        ${'OCONV ss ß'}   | ${{ ioConvertCost: 25 }} | ${{ map: '(ss)(ß)', replace: 25 }}
    `('affRepConv "$line" $costs', ({ line, costs, expected }) => {
        expect(affRepConv(line, calcCosts(costs))).toEqual(expected);
    });

    const AFFs: [string, HunspellCosts | undefined][] = [
        ['', undefined],
        [
            `
            TRY abc
            `,
            undefined,
        ],
        [sampleAff, undefined],
    ];

    test.each(AFFs)('hunspellInformationToSuggestionCostDef <%s>', (aff, costs) => {
        const info: HunspellInformation = { aff };
        if (costs) {
            info.costs = costs;
        }
        const defs = hunspellInformationToSuggestionCostDef(info);
        expect(defs).toMatchSnapshot();
    });
});

function c(hc: HunspellCosts = {}): HunspellCosts {
    return hc;
}
