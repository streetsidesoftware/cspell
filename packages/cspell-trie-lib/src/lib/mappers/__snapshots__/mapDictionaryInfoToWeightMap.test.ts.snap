// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapAffToWeightMap hunspellInformationToSuggestionCostDef <
            TRY abc
            > 1`] = `
Array [
  Object {
    "insDel": 95,
    "map": "abc",
    "replace": 95,
    "swap": 95,
  },
  Object {
    "map": "(^a)(^b)(^c)",
    "penalty": 4,
    "replace": 91,
  },
]
`;

exports[`mapAffToWeightMap hunspellInformationToSuggestionCostDef <
# comment.
# cspell:ignore eéèëê iíìïî
TRY abc

MAP 3
MAP eéèëê
MAP aáà
MAP iíìïî

OCONV 2
OCONV ĳ ij
OCONV Ĳ IJ

ICONV 9
ICONV áá aa
ICONV éé ee
ICONV íé ie
ICONV óó oo
ICONV úú uu
ICONV óé oe
ICONV ’ '
ICONV ij ĳ
ICONV IJ Ĳ

REP ^ß ss

NO-TRY -1234567890

# cspell:ignore asdfghjkl qwertyuiop zxcvbnm
KEY qwertyuiop|asdfghjkl|zxcvbnm
> 1`] = `
Array [
  Object {
    "insDel": 95,
    "map": "abc",
    "replace": 95,
    "swap": 95,
  },
  Object {
    "map": "(^a)(^b)(^c)",
    "penalty": 4,
    "replace": 91,
  },
  Object {
    "map": "eéèëê",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "aáà",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "iíìïî",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "(ĳ)(ij)",
    "replace": 30,
  },
  Object {
    "map": "(Ĳ)(IJ)",
    "replace": 30,
  },
  Object {
    "map": "(áá)(aa)",
    "replace": 30,
  },
  Object {
    "map": "(éé)(ee)",
    "replace": 30,
  },
  Object {
    "map": "(íé)(ie)",
    "replace": 30,
  },
  Object {
    "map": "(óó)(oo)",
    "replace": 30,
  },
  Object {
    "map": "(úú)(uu)",
    "replace": 30,
  },
  Object {
    "map": "(óé)(oe)",
    "replace": 30,
  },
  Object {
    "map": "(’)(')",
    "replace": 30,
  },
  Object {
    "map": "(ij)(ĳ)",
    "replace": 30,
  },
  Object {
    "map": "(IJ)(Ĳ)",
    "replace": 30,
  },
  Object {
    "map": "(^ß)(^ss)",
    "replace": 75,
  },
  Object {
    "insDel": 5,
    "map": "-1234567890",
    "penalty": 195,
  },
  Object {
    "map": "qw|we|er|rt|ty|yu|ui|io|op|as|sd|df|fg|gh|hj|jk|kl|zx|xc|cv|vb|bn|nm",
    "replace": 94,
    "swap": 94,
  },
]
`;

exports[`mapAffToWeightMap hunspellInformationToSuggestionCostDef <> 1`] = `Array []`;
